lib/SPIRV/SPIRVLowerSaddWithOverflow.cpp

```diff
@@ -0,0 +1,156 @@
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// Redistributions of source code must retain the above copyright notice,
+// this list of conditions and the following disclaimers.
+// Redistributions in binary form must reproduce the above copyright notice,
+// this list of conditions and the following disclaimers in the documentation
+// and/or other materials provided with the distribution.
+// Neither the names of Intel Corporation, nor the names of its
+// contributors may be used to endorse or promote products derived from this
+// Software without specific prior written permission.
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// CONTRIBUTORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS WITH
+// THE SOFTWARE.
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements lowering of llvm.sadd.with.overflow.* into basic LLVM
+// operations. Probably, in the future this pass can be generalized for other
+// function calls
+//
+//===----------------------------------------------------------------------===//
+#define DEBUG_TYPE "spv-lower-llvm_sadd_with_overflow"
+
+#include "LLVMSPIRVLib.h"
+#include "LLVMSaddWithOverflow.h"
+#include "SPIRVError.h"
+#include "llvm/IR/InstVisitor.h"
+#include "llvm/IR/IntrinsicInst.h"
+#include "llvm/IR/Module.h"
+#include "llvm/IR/Verifier.h"
+#include "llvm/IRReader/IRReader.h"
+#include "llvm/Linker/Linker.h"
+#include "llvm/Pass.h"
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/Debug.h"
+
+using namespace llvm;
+using namespace SPIRV;
+
+namespace SPIRV {
+cl::opt<bool> SPIRVLowerSaddWithOverflowValidate(
+    "spv-lower-saddwithoverflow-validate",
+    cl::desc("Validate module after lowering llvm.sadd.with.overflow.*"
+             "intrinsics"));
+
+class SPIRVLowerSaddWithOverflow
+    : public ModulePass,
+      public InstVisitor<SPIRVLowerSaddWithOverflow> {
+public:
+  SPIRVLowerSaddWithOverflow() : ModulePass(ID), Context(nullptr) {
+    initializeSPIRVLowerSaddWithOverflowPass(*PassRegistry::getPassRegistry());
+  }
+  virtual void visitIntrinsicInst(CallInst &I) {
+    IntrinsicInst *II = dyn_cast<IntrinsicInst>(&I);
+    if (!II || II->getIntrinsicID() != Intrinsic::sadd_with_overflow)
+      return;
+
+    Function *IntrinsicFunc = I.getCalledFunction();
+    assert(IntrinsicFunc && "Missing function");
+    StringRef IntrinsicName = IntrinsicFunc->getName();
+    std::string FuncName = "llvm_sadd_with_overflow_i";
+    if (IntrinsicName.endswith(".i16"))
```

Comment:
Minor detail: the commit message says you're only handling i32, but you seem to handle i16/i32/i64.

Reply:
Fixed.